### DISCUSSION TITLE
Is this a good solution to when updating nested objects?

### DISCUSSION THREAD

**Discussion by MarkWilliamsDev** [Automattic/mongoose#9872] [Show and tell]
I've been struggling all evening with the issue where updating one key in a nested object causes all values to be overwritten.  After loads of research looking for a simple solution, I tried something out where, as far as I understand it, any keys that don't receive a new value will update the new instance with the existing data.

     `const {
      name,
      venueContact,
      line1,
      line2,
      city,
      county,
      postCode,
      country,
      _id
    } = req.body

    try {
      const venue = await Venue.findById(_id)
      venue.name = name
      venue.venueContact = venueContact
      venue.address.line1 = line1 || venue.address.line1
      venue.address.line2 = line2 || venue.address.line2
      venue.address.city = city || venue.address.city
      venue.address.county = county || venue.address.county
      venue.address.postCode = postCode || venue.address.postCode
      venue.address.country = country || venue.address.country
      venue.save()

      res.send(venue)
    } catch (err) {

      res.send(err)
    }`
      
It seems to work but am I missing something that could become an issue?

**Comment by AndrewFriendTech** [comment]
If you have a rapidly updating database, you could cause race conditions where multiple clients are updating the same code at once and end up with garbled objects if one user submits the change after another user has downloaded the object, but before his client has submitted it. 

For your situation , it is probally best to use Venue.Update() queries. See [here](https://mongoosejs.com/docs/documents.html#updating-using-queries) in the mongoose docs.

Sorry for replying so late to your post,  i don't normally browse the discussions of here, you might find stack overflow will give you faster answer. (but they can tend to be rude at times as well)

**Reply by AbdelrahmanHafez** [reply]
@AndrewFriendTech `Document#save(...)` uses `Model.updateOne(...)`/`Model.insertOne(...)` under the hood, you have a point regarding the race condition though, he needs to `await` on `venu.save()` to avoid _some_ race-conditions.

**Comment by AbdelrahmanHafez** [comment]
That does work, but that seems to be too much repetition and that's an anti-pattern.
You seem to want to set an item if it's present in `req.body`.
A different approach would be to build a `clean` utility that removes any `null` or `undefined` properties from `req.body` and `document.set(req.body);` then `document.save();`
```js
try {
  const venue = await Venue.findById(_id);

  clean(req.body);
  revenue.set(req.body);

  await venue.save();

  return  res.send(venue);
} catch (err) {
  res.send(err)
}

function clean(obj) {
  for (const propName in obj) {
    // technically, you'd need to do `!obj[propName]` to have the equivalent implementation
   // of what you have in the original post
    if (obj[propName] == null) { 
      delete obj[propName];
    }
  }
```

Also, as @AndrewFriendTech has mentioned, you want to `await venue.save();` to be over, so that if an error happened, you'd notify the user with it instead of assuming that the `save` will just work.

One more thing I'd do with the changes above is validate the `req.body` using [joi](https://joi.dev/api/) to avoid malicious users sending more information (for example `venue.size`) than intended which will eventually go to the database.
